core/src/main/java/org/apache/iceberg/rest/http/HttpClient.java

```diff
@@ -0,0 +1,235 @@
+ */
+public class HttpClient implements RestClient {
+  private final String baseUrl;
+
+  // Need their own Builders?
+  private final CloseableHttpClient httpClient;
+  private final Consumer<CloseableHttpResponse> defaultErrorHandler;
+  private final ObjectMapper mapper;
+  private Map<String, String> additionalHeaders = Maps.newHashMap();
+  private Consumer<HttpUriRequest> requestInterceptor = (r) -> { };
+
+  private HttpClient(
+      String baseUrl, CloseableHttpClient httpClient, ObjectMapper mapper,
+      Map<String, String> additionalHeaders, Consumer<HttpUriRequest> requestIntercepter,
+      Consumer<CloseableHttpResponse> defaultErrorHandler) {
+    this.baseUrl = baseUrl;
+    this.httpClient = httpClient != null ? httpClient : HttpClients.createDefault();
+    this.mapper = mapper != null ? mapper : new ObjectMapper();
+    this.additionalHeaders = additionalHeaders != null ? additionalHeaders : Maps.newHashMap();
+    // TODO - Best way to handle Preconditions here.
+    this.requestInterceptor = requestIntercepter;
+    this.defaultErrorHandler = defaultErrorHandler;
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  /**
+   * Method to execute an HTTP request and process the corresponding response.
+   *
+   * @param method       - HTTP method, such as GET, POST, HEAD, etc.
+   * @param path         - URL path to send the request to
+   * @param body         - Contents of the request body.
+   * @param responseType - Class of the Response type. Needs to have serializer registered with ObjectMapper
+   * @param errorHandler - Error handler delegated for HTTP responses which handles server error responses
+   * @param <T>          - Class type of the response for deserialization. Must be registered with the ObjectMapper.
+   * @return The Response enttity, parsed and converted to its type T
+   * @throws UncheckedIOException - Shouldn't throw this as the requestInterceptor should handle expected cases.
+   * @throws UncheckedRestException - Wraps exceptions to avoid having to use checked exceptions everywhere.
+   */
+  @Nullable
+  public <T> T execute(
+      Method method, String path, Object body, Class<T> responseType,
+      Consumer<CloseableHttpResponse> errorHandler) {
+    HttpUriRequestBase request = new HttpUriRequestBase(method.name(), URI.create(String.format("%s/%s", baseUrl,
+        path)));
+    addRequestHeaders(request);
+
+    if (body != null) {
+      try {
+        request.setEntity(new StringEntity(mapper.writeValueAsString(body)));
+      } catch (JsonProcessingException e) {
+        throw new UncheckedRestException(e, "Failed to write request body");
+      }
+    }
+
+    requestInterceptor.accept(request);
+
+    try (CloseableHttpResponse response = httpClient.execute(request)) {
+      if (response.getCode() != HttpStatus.SC_OK) {
+        errorHandler.accept(response);
```

Comment:
Should this return after handling the response in case it doesn't throw an exception?

Reply:
+1. I decided to use the Apache HTTP library, as we're in an Apache project. But I'm more familiar with other HTTP libraries. I will return here and then consider more about libraries etc. I'm a big fan of OkHTTP, but I recognize that might not be in-line with the fact that this is an Apache repo.

Reply:
We can use whatever we want as long as the license is compatible. That said, if the library is already included with Spark and Flink, we generally prefer to use it so we don't need to shade and relocate it. That's why Apache HTTP is most often used.